shuup/core/api/basket.py

```diff
@@ -297,42 +296,71 @@ def _handle_cmd(self, request, command, kwargs):
+        if "shop_product" in data:
+            serializer = ShopProductAddBasketSerializer(data=data, context={"shop": self.request.shop})
         else:
-            serializer = ProductAddBasketSerializer(data=request.data, context={"shop": request.shop})
+            serializer = ProductAddBasketSerializer(data=data, context={"shop": self.request.shop})
 
-        if serializer.is_valid():
+        if serializer.is_valid(raise_exception=True):
             cmd_kwargs = {
-                "request": request._request,
-                "basket": request._request.basket,
+                "request": self.request._request,
+                "basket": self.request._request.basket,
                 "shop_id": serializer.data.get("shop") or serializer.validated_data["shop"].pk,
                 "product_id": serializer.data.get("product") or serializer.validated_data["product"].pk,
                 "quantity": serializer.validated_data.get("quantity", 1),
                 "supplier_id": serializer.validated_data.get("supplier")
             }
-            # we call `add` directly, assuming the user will handle variations
-            # as he can know all product variations easily through product API
+            self._handle_cmd(self.request._request, "add", cmd_kwargs)
+            self.request.basket.save()
+
+    @detail_route(methods=['post'])
+    def add(self, request, *args, **kwargs):
+        """
+        Adds a product to the basket
+        """
+        self.process_request()
+        try:
+            self.add_product(request.data)
+        except ValidationError as exc:
+            return Response({exc.code: exc.message}, status=status.HTTP_400_BAD_REQUEST)
+        except ProductNotOrderableProblem as exc:
+            return Response({"error": "{}".format(exc)}, status=status.HTTP_400_BAD_REQUEST)
+        except serializers.ValidationError as exc:
+            return Response(exc.get_full_details(), status=status.HTTP_400_BAD_REQUEST)
+        except Exception as exc:
+            return Response({"error": str(exc)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+        else:
+            return Response(BasketSerializer(request.basket, context=self.get_serializer_context()).data)
+
+    @detail_route(methods=['post'])
+    def add_from_order(self, request, *args, **kwargs):
+        """
+        Add multiple products to the basket
+        """
+        self.process_request()
+        errors = []
+        order = Order.objects.filter(
+            customer=get_person_contact(request.user),
+            pk=request.data.get("order")
+        ).first()
+        if not order:
+            return Response({"error": "invalid order"}, status=status.HTTP_404_NOT_FOUND)
+        for line in order.lines.products():
             try:
-                self._handle_cmd(request, "add", cmd_kwargs)
-                request.basket.save()
+                self.add_product({"product": line.product_id, "shop": order.shop_id, "quantity": line.quantity})
```

Comment:
what happens if user has already something in the basket? should we clean it first?